File: adaptive_repair/data/python_programs/longest_common_subsequence.py
def longest_common_subsequence(a, b):
    # Use a dictionary to store results of subproblems (memoization table)
    # Keys will be tuples (index_a, index_b) representing the starting indices
    # of the substrings being considered (a[index_a:] and b[index_b:]).
    # Values will be the LCS string for those substrings.
    memo = {}

    def _lcs_recursive(i, j):
        # If this subproblem has already been solved, return the stored result
        if (i, j) in memo:
            return memo[(i, j)]

        # Base case: If either string (from its current index) is exhausted,
        # there's no common subsequence possible from this point onward.
        if i == len(a) or j == len(b):
            return ''

        # If the current characters at indices i and j match
        if a[i] == b[j]:
            # They are part of the LCS. Include a[i] (or b[j]) and
            # recursively find the LCS of the remaining parts of the strings.
            result = a[i] + _lcs_recursive(i + 1, j + 1)
        else:
            # If characters don't match, we have two choices:
            # 1. Skip the character from string 'a' (a[i]) and find LCS of a[i+1:] and b[j:]
            # 2. Skip the character from string 'b' (b[j]) and find LCS of a[i:] and b[j+1:]
            # We take the longer of the two results.
            result = max(
                _lcs_recursive(i, j + 1),    # Skip b[j]
                _lcs_recursive(i + 1, j),    # Skip a[i]
                key=len
            )

        # Store the computed result in the memoization table before returning
        # to avoid re-computation if this subproblem is encountered again.
        memo[(i, j)] = result
        return result

    # Start the recursive process from the beginning of both strings (indices 0, 0)
    return _lcs_recursive(0, 0)


"""
Longest Common Subsequence


Calculates the longest subsequence common to the two input strings. (A subsequence is any sequence of letters in the same order
they appear in the string, possibly skipping letters in between.)

Input:
    a: The first string to consider.
    b: The second string to consider.

Output:
    The longest string which is a subsequence of both strings. (If multiple subsequences of equal length exist, either is OK.)

Example:
    >>> longest_common_subsequence('headache', 'pentadactyl')
    'eadac'
"""

File: adaptive_repair/data/java_programs/LONGEST_COMMON_SUBSEQUENCE.java
package java_programs;

public class LONGEST_COMMON_SUBSEQUENCE {

    private static String[][] memo;
    private static String strA;
    private static String strB;
    private static int lenA;
    private static int lenB;

    public static String longest_common_subsequence(String a, String b) {
        strA = a;
        strB = b;
        lenA = a.length();
        lenB = b.length();

        // Initialize memoization table.
        // The size is (lenA + 1) x (lenB + 1) because indices go up to lenA and lenB,
        // and we need to store results for (lenA, j) and (i, lenB) as base cases.
        memo = new String[lenA + 1][lenB + 1];

        return _lcs(0, 0);
    }

    private static String _lcs(int i, int j) {
        // Check if the result for this subproblem (i, j) has already been computed.
        if (memo[i][j] != null) {
            return memo[i][j];
        }

        // Base case: if either string segment is exhausted, there's no common subsequence.
        // Return an empty string.
        if (i == lenA || j == lenB) {
            return "";
        }

        String result;
        // If the current characters from both strings match, they are part of the LCS.
        // Append the character and recurse for the rest of the strings (i+1, j+1).
        if (strA.charAt(i) == strB.charAt(j)) {
            result = strA.charAt(i) + _lcs(i + 1, j + 1);
        } else {
            // If characters don't match, explore two possibilities:
            // 1. Skip the current character in string B (move to B[j+1])
            // 2. Skip the current character in string A (move to A[i+1])
            // Take the longer of the two resulting LCSs.
            String lcs1 = _lcs(i, j + 1); // LCS when A[i] is considered, but B[j] is skipped
            String lcs2 = _lcs(i + 1, j); // LCS when B[j] is considered, but A[i] is skipped
            result = (lcs1.length() >= lcs2.length()) ? lcs1 : lcs2;
        }

        // Store the computed result in the memoization table before returning,
        // so it can be reused if this subproblem is encountered again.
        memo[i][j] = result;
        return result;
    }
}